x-pack/filebeat/input/salesforce/input.go

```diff
@@ -561,22 +589,21 @@ func decodeAsCSV(p []byte) ([]map[string]string, error) {
 	// so that future records must have the same field count.
 	// So, if len(header) != len(event), the Read will return an error and hence
 	// we need not put an explicit check.
-	event, err := r.Read()
-	for ; err == nil; event, err = r.Read() {
+	for {
+		record, err := r.Read()
 		if err != nil {
-			continue
-		}
-		o := make(map[string]string, len(header))
-		for i, h := range header {
-			o[h] = event[i]
+			if errors.Is(err, io.EOF) {
+				break
+			}
+			s.log.Errorf("failed to read CSV record: %v\n%s", err, p)
+			return nil, textContextError{error: fmt.Errorf("failed to read CSV record: %w for: %v", err, record), body: p}
 		}
-		results = append(results, o)
-	}
 
-	if err != nil {
-		if err != io.EOF { //nolint:errorlint // csv.Reader never wraps io.EOF.
-			return nil, textContextError{error: err, body: p}
+		event := make(map[string]string, len(header))
```

Comment:
For later, since we are reallocating a map here, we could try the sync.Pool(https://pkg.go.dev/sync#Pool) construct to reduce the allocation overhead and recycle/reuse the map. If we see some gains, we should definitely replace it with a Pool.

Reply:
sync.Pool can be used but the map shouldn't be reused. Check here: https://github.com/elastic/beats/pull/40904/files; we recently faced this. During publishing some process is there where the goroutine concurrently operates on the map. So in this PR, the map was being reused as it was constant for all and that caused the panic.

I'll keep a note of sync.Pool but I have to test with the locking and all; in terms of CPU is it similar or worse in this particular case because for concurrent code it does make sense. I'll try it. Thanks!

Btw, one optimization we already have done in this CSV unmarshaler is the `r.ReuseRecord = true`.